ports/mimxrt/machine_i2c.c

```diff
@@ -0,0 +1,217 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "extmod/machine_i2c.h"
+#include "modmachine.h"
+
+#include "fsl_iomuxc.h"
+#include "fsl_lpi2c.h"
+
+#define DEFAULT_I2C_FREQ        (400000)
+#define DEFAULT_I2C_DRIVE       (6)
+
+// Select USB1 PLL (480 MHz) as master lpi2c clock source
+#define LPI2C_CLOCK_SOURCE_SELECT (0U)
+// Clock divider for master lpi2c clock source
+#define LPI2C_CLOCK_SOURCE_DIVIDER (1U)
+// Get frequency of lpi2c clock = 30 MHz
+#define LPI2C_CLOCK_FREQUENCY ((CLOCK_GetFreq(kCLOCK_Usb1PllClk) / 8) / (LPI2C_CLOCK_SOURCE_DIVIDER + 1U))
+
+typedef struct _machine_i2c_obj_t {
+    mp_obj_base_t base;
+    LPI2C_Type *i2c_inst;
+    uint8_t i2c_id;
+    uint8_t i2c_hw_id;
+    bool transfer_busy;
+    status_t transfer_status;
+    lpi2c_master_config_t *master_config;
+} machine_i2c_obj_t;
+
+typedef struct _iomux_table_t {
+    uint32_t muxRegister;
+    uint32_t muxMode;
+    uint32_t inputRegister;
+    uint32_t inputDaisy;
+    uint32_t configRegister;
+} iomux_table_t;
+
+STATIC const uint8_t i2c_index_table[] = MICROPY_HW_I2C_INDEX;
+STATIC LPI2C_Type *i2c_base_ptr_table[] = LPI2C_BASE_PTRS;
+static const iomux_table_t iomux_table[] = { IOMUX_TABLE_I2C };
+
+#define MICROPY_HW_I2C_NUM     ARRAY_SIZE(i2c_index_table)
+
+#define SCL (iomux_table[index])
+#define SDA (iomux_table[index + 1])
+
+bool lpi2c_set_iomux(int8_t hw_i2c, uint8_t drive) {
+    int index = (hw_i2c - 1) * 2;
+
+    if (SCL.muxRegister != 0) {
+        IOMUXC_SetPinMux(SCL.muxRegister, SCL.muxMode, SCL.inputRegister, SCL.inputDaisy, SCL.configRegister, 1U);
+        IOMUXC_SetPinConfig(SCL.muxRegister, SCL.muxMode, SCL.inputRegister, SCL.inputDaisy, SCL.configRegister,
+            0xF880u | drive << IOMUXC_SW_PAD_CTL_PAD_DSE_SHIFT);
+
+        IOMUXC_SetPinMux(SDA.muxRegister, SDA.muxMode, SDA.inputRegister, SDA.inputDaisy, SDA.configRegister, 1U);
+        IOMUXC_SetPinConfig(SDA.muxRegister, SDA.muxMode, SDA.inputRegister, SDA.inputDaisy, SDA.configRegister,
+            0xF880u | drive << IOMUXC_SW_PAD_CTL_PAD_DSE_SHIFT);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+STATIC void machine_i2c_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_i2c_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "I2C(%u, freq=%u)",
+        self->i2c_id, self->master_config->baudRate_Hz);
+}
+
+mp_obj_t machine_i2c_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_id, ARG_freq, ARG_drive};
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_freq, MP_ARG_INT, {.u_int = DEFAULT_I2C_FREQ} },
+        { MP_QSTR_drive, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_I2C_DRIVE} },
```

Comment:
what arguments can be passed to `drive`?

Reply:
The same as for the Pin module and and the SPI module. Values 0 - 6 or Pin.POWER_0 to Pin.POWER_6. It changes the hardware pin impedance.  Since pins are assigned fixed to the I2C id's, not Pin settings can be made. But obviously this is only critical for the low edge. And since the I2C speed is low, it is less important than for SPI.